File: Lab04/cpp/main.cpp
#include <iostream>
#include <cmath>
#include <iomanip>
#include <thread>
#include <chrono>

#include "CalculadorIntegral.h"
#include "Funcion.h"

int main() {
    const double A = 2.0;
    const double B = 20.0;
    const int NUM_HILOS = std::thread::hardware_concurrency();

    Funcion f;
    CalculadorIntegral calculador(f, A, B);

    // --- Mensajito inicial mejorado ---
    std::cout << "+ Iniciando calculo con la siguiente configuracion:" << std::endl;
    std::cout << "- Funcion: f(x) = 2x^2 + 3x + 0.5" << std::endl;
    std::cout << "- Limites: [" << A << ", " << B << "]" << std::endl;
    std::cout << "- Hilos detectados: " << NUM_HILOS << std::endl;
    std::cout << "--------------------------------------------------" << std::endl;

    // --- INICIO DE LA MEDICIÓN DE TIEMPO ---
    auto start = std::chrono::high_resolution_clock::now();

    double area_anterior = 0.0;
    double area_actual = -1.0;
    int n_trapecios = 1;
    const double TOLERANCIA = 1e-7; // Criterio de parada

    std::cout << "Calculando..." << std::endl;
    std::cout << std::fixed << std::setprecision(8);

    // En esta parte hago que el bucle se detenga cuando el resultado ya no cambia significativamente
    while (std::abs(area_actual - area_anterior) > TOLERANCIA) {
        area_anterior = area_actual;
        // Cada llamada a 'calcular' ahora ejecutará el trabajo en paralelo.
        area_actual = calculador.calcular(n_trapecios);
        std::cout << "Trapecios: " << n_trapecios << ", Area: " << area_actual << std::endl;

        // Para mejorar la convergencia se duplica
        n_trapecios *= 2;
    }

    // --- FIN PARA LA MEDICIÓN DE TIEMPO ---
    auto end = std::chrono::high_resolution_clock::now();

    // Duración total en milisegundos.
    auto duration_ms = std::chrono::duration_cast<std::chrono::milliseconds>(end - start).count();


    // --- RESUMEN FINAL ---
    std::cout << "\n--- RESUMEN DE EJECUCION ---" << std::endl;
    std::cout << "Resultado final estable." << std::endl;
    std::cout << "- Area calculada:      " << area_actual << std::endl;
    std::cout << "- Trapecios usados:     " << n_trapecios / 2 << std::endl; // El último valor se duplicó, así que lo dividimos.
    std::cout << "- Hilos de ejecucion:   " << NUM_HILOS << std::endl;
    std::cout << "- Tiempo total:         " << duration_ms << " ms" << std::endl;
    std::cout << "--------------------------" << std::endl;

    return 0;
}

File: Lab04/go/main.go
package main

import (
	"fmt"
	"math"
	"runtime"
	"time"
)

func main() {
	const A = 2.0
	const B = 20.0
	NUM_GOROUTINES := runtime.NumCPU()

	fmt.Println(" GOLANG - Calculadora paralela")
	fmt.Println("=============================")
	fmt.Printf("Funcion a integrar: f(x) = 2x^2 + 3x + 0.5\n")
	fmt.Printf("Limites: [%.1f, %.1f]\n", A, B)
	fmt.Printf("Goroutines a utilizar (nucleos logicos): %d\n", NUM_GOROUTINES)
	fmt.Println("=============================")

	start := time.Now()

	areaAnterior := 0.0
	areaActual := -1.0
	nTrapecios := 1
	const TOLERANCIA = 1e-7

	fmt.Println("\nIniciando convergencia...")
	for math.Abs(areaActual-areaAnterior) > TOLERANCIA {
		areaAnterior = areaActual
		areaActual = calcularIntegralParalelo(miFuncion, A, B, nTrapecios, NUM_GOROUTINES)

		// Usamos '\r' (retorno de carro) para que el cursor vuelva al inicio de la línea
		fmt.Printf("\r -> Iterando... Trapecios: %-10d | Area actual: %.8f", nTrapecios, areaActual)

		nTrapecios *= 2
	}

	duration := time.Since(start)

	fmt.Println()

	fmt.Println("\n//======= RESULTADOS FINALES =======//")
	fmt.Println("Convergencia alcanzada.")
	fmt.Printf("Area final:\t\t%.8f\n", areaActual)
	fmt.Printf("Trapecios necesarios:\t%d\n", nTrapecios/2)
	fmt.Printf("Goroutines concurrentes:\t%d\n", NUM_GOROUTINES)
	fmt.Printf("Tiempo de ejecucion:\t%d ms\n", duration.Milliseconds())
	fmt.Println("//==================================//")
}
